File: iRank Timer/ConfigMasterViewController.m
//
//  ConfigMasterViewController.m
//  iRank Timer
//
//  Created by Luciano Stegun on 2013-09-29.
//  Copyright (c) 2013 Stegun.com. All rights reserved.
//

#import "ConfigMasterViewController.h"
#import "ConfigViewController.h"
#import "ConfigLanguageViewController.h"

@interface ConfigMasterViewController ()

@end

@implementation ConfigMasterViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
 
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
    
    switchPlaySounds.on       = appDelegate.playSounds;
    switchFiveSecondsClock.on = appDelegate.fiveSecondsClock;
    switchLongPauseAlert.on   = appDelegate.longPauseAlert;
    switchShortBlindNumber.on = appDelegate.shortBlindNumber;
    switchPreventSleep.on     = appDelegate.preventSleep;
    switchRunBackground.on    = appDelegate.runBackground;
    
    [self localizeView];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(changeMasterDetails:) name:kNotificationSettingsDetailsChange object:nil];
}

-(void)localizeView {
    
    self.title = NSLocalizedString(@"iRank Timer Config", "ConfigMasterViewController");

    lblPlaySounds.text       = NSLocalizedString(@"Play sounds", "ConfigMasterViewController");
    lblSounds.text           = NSLocalizedString(@"Sounds", "ConfigMasterViewController");
    lblFiveSecondsAlert.text = NSLocalizedString(@"FiveSecondsAlert", "ConfigMasterViewController");
    lblLongPauseAlert.text   = NSLocalizedString(@"LongPauseAlert", "ConfigMasterViewController");
    lblShortBlindNumber.text = NSLocalizedString(@"ShortBlindNumber", "ConfigMasterViewController");
    lblLanguageLabel.text    = NSLocalizedString(@"Language", "ConfigMasterViewController");
    lblPreventSleep.text     = NSLocalizedString(@"PreventSleep", "ConfigMasterViewController");
    lblRunBackground.text    = NSLocalizedString(@"RunBackground", "ConfigMasterViewController");
    
    [self changeMasterDetails:nil];
}

- (void)changeMasterDetails:(NSNotification *)notification {
    
    lblBlindChangeSound.text = [self blindChangeSoundName];
    lblLanguage.text         = [self language];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 3;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{

    switch( section ){
        case 0:
            if( switchPlaySounds.on )
                return 5;
            else
                return 1;
        case 1:
            return 1;
        case 2:
            return 2;
    }
    return 0;
}

-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    
    switch( section ){
        case 0:
            return NSLocalizedString(@"Sound settings", "ConfigMasterViewController");
        case 1:
            return nil;
        case 2:
            return NSLocalizedString(@"General", "ConfigMasterViewController");
    }
    
    return nil;    
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {

//    ConfigViewController *configViewController = (ConfigViewController*)self.parentViewController.parentViewController;
//
//
    if( indexPath.section==0 ){

        ConfigDetailViewController *configDetailViewController = [[ConfigDetailViewController alloc] initWithStyle:UITableViewStyleGrouped];

//        self.detailViewController.detailItem = configDetailViewController;
        
//        [(UINavigationController*)self.detailViewController.parentViewController setViewControllers:[NSArray arrayWithObject:configDetailViewController]];

//        configViewController.viewControllers = [NSArray arrayWithObjects:self.parentViewController, configDetailViewController, nil];
        
        
        [self changeDetailViewController:configDetailViewController];
//        [[[(UISplitViewController*)self.parentViewController.parentViewController viewControllers] objectAtIndex:1] setViewControllers:[NSArray arrayWithObject:configDetailViewController]];
    }else{

        ConfigLanguageViewController *configLanguageViewController = [[ConfigLanguageViewController alloc] initWithStyle:UITableViewStyleGrouped];
//        [[configViewController.viewControllers objectAtIndex:1] pushViewController:configLanguageViewController];
//        configViewController.viewControllers = [NSArray arrayWithObjects:self.parentViewController, configLanguageViewController, nil];
        
//        self.detailViewController.detailItem = configLanguageViewController;
        
//        [(UINavigationController*)self.detailViewController.parentViewController setViewControllers:[NSArray arrayWithObject:configLanguageViewController]];
        
        [self changeDetailViewController:configLanguageViewController];
    }
    
//    if (self.detailViewController.masterPopoverController != nil)
//        [self.detailViewController.masterPopoverController dismissPopoverAnimated:YES];
}

-(void)changeDetailViewController:(UIViewController*)viewController {
    
    [[[(UISplitViewController*)self.parentViewController.parentViewController viewControllers] objectAtIndex:1] setViewControllers:[NSArray arrayWithObject:viewController]];
}

-(void)saveSettings:(id)sender {
    
    appDelegate.playSounds       = switchPlaySounds.on;
    appDelegate.fiveSecondsClock = switchFiveSecondsClock.on;
    appDelegate.longPauseAlert   = switchLongPauseAlert.on;
    appDelegate.shortBlindNumber = switchShortBlindNumber.on;
    appDelegate.preventSleep     = switchPreventSleep.on;
    appDelegate.runBackground    = switchRunBackground.on;
    [appDelegate updateSettings];
    
    [self dismiss];
}

-(void)dismiss {
    
    [appDelegate dismissRootViewController];
}

-(NSString*)blindChangeSoundName {
    
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"blindChangeSounds" ofType:@"plist"];
    NSArray *blindChangeSoundList = [[NSArray alloc] initWithContentsOfFile:filePath];
    
    for(NSDictionary *item in blindChangeSoundList){
        
        if( [appDelegate.blindChangeSound isEqualToString:item[@"fileName"]] ){
         
            blindChangeSoundList = nil;
            return item[@"soundName"];
        }
    }
    
    blindChangeSoundList = nil;
    return nil;
}

-(NSString*)language {
    
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"languages" ofType:@"plist"];
    NSArray *languageList = [[NSArray alloc] initWithContentsOfFile:filePath];
    
    for(NSDictionary *menuItem in languageList ){
        
        if( [appDelegate.language isEqualToString:menuItem[@"culture"]] ){

            languageList = nil;
            return menuItem[@"language"];
        }
    }
    
    languageList = nil;
    return nil;
}

-(void)switchSoundSettings:(id)sender {
    
    [self.tableView reloadData];
}

@end

File: iRank Timer/ConfigMasterViewController.swift
//
//  ConfigMasterViewController.swift
//  iRank Timer
//
//  Created by Luciano Stegun on 11/02/15.
//  Copyright (c) 2015 Stegun.com. All rights reserved.
//

import UIKit
import Foundation

class ConfigMasterViewController : UITableViewController {
    
    let appDelegate = UIApplication.sharedApplication().delegate as! AppDelegate;
    var firstAppear = true;
    
    @IBOutlet var swcPlaySounds : UISwitch!;
    @IBOutlet var swcFiveSecondsClock : UISwitch!;
    @IBOutlet var swcPreventSleep : UISwitch!;
    @IBOutlet weak var swcEnableMultipeer: UISwitch!
    @IBOutlet weak var swcBackgroundRun: UISwitch!
    @IBOutlet weak var swcShortBlindNumber: UISwitch!
    @IBOutlet weak var swcInverseTextColor: UISwitch!
    @IBOutlet weak var swcHideStatusBar: UISwitch!
    @IBOutlet weak var lblSoundsName: UILabel!
    @IBOutlet weak var lblLongPauseMinutes: UILabel!
    @IBOutlet weak var lblLanguage: UILabel!
    @IBOutlet weak var lblEnableBackup: UILabel!
    @IBOutlet weak var lblBackgroundMode: UILabel!
    @IBOutlet weak var lblMultipeerDevice: UILabel!
    
    var configDetailsMultipeerViewController : UITableViewController!;
    
    override func viewDidLoad() {

        updateLabels();

        NSNotificationCenter.defaultCenter().addObserver(self, selector: Selector("updateLabels"), name: "changeSubOptionValue", object: nil);
        
        swcPlaySounds.on       = appDelegate.playSounds;
        swcFiveSecondsClock.on = appDelegate.fiveSecondsClock;
        swcPreventSleep.on     = appDelegate.preventSleep;
        swcShortBlindNumber.on = appDelegate.shortBlindNumber;
        swcInverseTextColor.on = appDelegate.inverseTextColor;
        swcHideStatusBar.on    = appDelegate.hideStatusBar;
        swcEnableMultipeer.on  = appDelegate.enableMultipeer;
        lblEnableBackup.text   = appDelegate.enableBackup ? NSLocalizedString("Enabled", comment: "") : NSLocalizedString("Disabled", comment: "");
    }
    
    override func prefersStatusBarHidden() -> Bool {
        return false;//appDelegate.hideStatusBar;
    }
    
    override func preferredStatusBarStyle() -> UIStatusBarStyle {
        
        return UIStatusBarStyle.LightContent;
    }
    
    override func supportedInterfaceOrientations() -> Int {
        
        if( Constants.DeviceIdiom.IS_IPAD ){
            
            return Int(UIInterfaceOrientationMask.LandscapeLeft.rawValue) | Int(UIInterfaceOrientationMask.LandscapeRight.rawValue)
        }
        
        return Int(UIInterfaceOrientationMask.All.rawValue);
    }
    
    override func viewWillAppear(animated: Bool) {

        super.viewWillAppear(animated);
        
        if( firstAppear && Constants.DeviceIdiom.IS_IPAD ){
            
            firstAppear = false;
        }
    }
    
//    func autoSelectMasterRow(){
//        
//        var indexPath : NSIndexPath!;
//        
//        if( swcPlaySounds.on ){
//            
//            indexPath = NSIndexPath(forRow: 1, inSection: 0);
//        }else{
//            
//            indexPath = NSIndexPath(forRow: 2, inSection: 1);
//        }
//        
//        self.tableView(self.tableView, didSelectRowAtIndexPath: indexPath);
//        self.tableView.selectRowAtIndexPath(indexPath, animated: false, scrollPosition: UITableViewScrollPosition.None);
//    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        switch( section ){
        case 0:
            return 4;
        case 1:
            return (Constants.DeviceIdiom.IS_IPHONE && Constants.LITE_VERSION ? 4 : 5);
        case 2:
            return (appDelegate.enableMultipeer && Constants.DeviceIdiom.IS_IPAD ? 5 : 4);
        default:
            return 0;
        }
    }
    
    @IBAction func togglePlaySounds(sender: AnyObject) {
        
        appDelegate.playSounds = swcPlaySounds.on;
        self.tableView.reloadData();
    }
    
    @IBAction func toggleFiveSecondsClock(sender: AnyObject) {
        
        appDelegate.fiveSecondsClock = swcFiveSecondsClock.on;
    }
    
    @IBAction func togglePreventSleep(sender: AnyObject) {

        appDelegate.preventSleep = swcPreventSleep.on;
    }
    
    @IBAction func toggleBackgroundRun(sender: AnyObject) {
        
        appDelegate.backgroundRun = swcBackgroundRun.on;
    }
    
    @IBAction func toggleShortBlindNumber(sender: AnyObject) {
        
        appDelegate.shortBlindNumber = swcShortBlindNumber.on;
    }
    
    @IBAction func toggleEnableMultipeer(sender: AnyObject) {
        
        if( !appDelegate.checkLiteVersion() ){
        
            appDelegate.enableMultipeer = false;
            swcEnableMultipeer.on       = false;
        }
        
        appDelegate.enableMultipeer = swcEnableMultipeer.on;
        self.tableView.reloadData();
        
        if( swcEnableMultipeer.on ){
            
            if( appDelegate.mpcManager == nil ){
                
                appDelegate.enableMultipeerConnection()
            }
            
            appDelegate.timerViewController.checkMultipeerMode();
        }else{
            
            appDelegate.mpcManager = nil;
        }
    }
    
    @IBAction func toggleInverseTextColor(sender: AnyObject) {
        
        appDelegate.inverseTextColor = swcInverseTextColor.on;
    }
    
    @IBAction func toggleHideStatusBar(sender: AnyObject) {
        
        appDelegate.hideStatusBar = swcHideStatusBar.on;
    }
    
    @IBAction func saveSettings(sender: AnyObject) {
        
        appDelegate.saveUserDefaults();
        self.dismissViewControllerAnimated(true, completion: nil);
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        
        switch( indexPath.section ){
        case 0:
            switch( indexPath.row ){
            case 1 ... 3:
                return swcPlaySounds.on ? 44: 0;
            default:
                break;
            }
            break;
        case 1:
            break;
        case 2:
            switch( indexPath.row ){
            case 4:
                return swcEnableMultipeer.on ? 44: 0;
            default:
                break;
            }
            break;
        default:
            break;
        }
        
        return 44;
    }

    override func tableView(tableView: UITableView, willDisplayCell cell: UITableViewCell, forRowAtIndexPath indexPath: NSIndexPath) {
        
        switch( indexPath.section ){
        case 0:
            switch( indexPath.row ){
            case 1 ... 3:
                return cell.hidden = !swcPlaySounds.on;
            default:
                break;
            }
            break;
        case 2:
            switch( indexPath.row ){
            case 4:
                return cell.hidden = !swcEnableMultipeer.on;
            default:
                break;
            }
            break;
        default:
            break;
        }
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        
        if( Constants.DeviceIdiom.IS_IPHONE ){
            
            return;
        }
        
        let storyboard = UIStoryboard(name: "iPad_Config", bundle: nil);
        
        switch( indexPath.section ){
        case 0:
            switch( indexPath.row ){
            case 1:
                let vc         = storyboard.instantiateViewControllerWithIdentifier("ConfigDetailsSoundViewController") as! UITableViewController;
                self.changeDetailViewController(vc);
            case 3:
                let vc         = storyboard.instantiateViewControllerWithIdentifier("ConfigDetailsLongPauseViewController") as! UITableViewController;
                self.changeDetailViewController(vc);
            default:
                break;
            }
            break;
        case 1:
            switch( indexPath.row ){
            case 0:
                let vc         = storyboard.instantiateViewControllerWithIdentifier("ConfigDetailsLanguageViewController") as! UITableViewController;
                self.changeDetailViewController(vc);
            case 2:
                let vc         = storyboard.instantiateViewControllerWithIdentifier("ConfigDetailsBackgroundViewController") as! UITableViewController;
                self.changeDetailViewController(vc);
            default:
                break;
            }
            break;
        case 2:
            switch( indexPath.row ){
            case 0:
                let vc         = storyboard.instantiateViewControllerWithIdentifier("ConfigDetailsBackupViewController") as! UITableViewController;
                self.changeDetailViewController(vc);
            case 4:
                if( configDetailsMultipeerViewController == nil ){
                    
                    configDetailsMultipeerViewController = storyboard.instantiateViewControllerWithIdentifier("ConfigDetailsMultipeerViewController") as! UITableViewController;
                }
                
                self.changeDetailViewController(configDetailsMultipeerViewController);
            default:
                break;
            }
            break;
        default:
            break;
        }
        
        tableView.selectRowAtIndexPath(indexPath, animated: false, scrollPosition: UITableViewScrollPosition.None);
    }
    
    override func tableView(tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        
        if( section == 2 ){
            
            return "www.irank.com.br\n\n© stegun.com";
        }
        
        return "";
    }
    
    func changeDetailViewController(viewController : UIViewController){
     
        if( Constants.DeviceIdiom.IS_IPAD ){
            
            var splitViewController = self.parentViewController?.parentViewController as! UISplitViewController;
            
            (splitViewController.viewControllers.last as! UINavigationController).setViewControllers([viewController], animated: false);
        }
    }
    
    func updateLabels(){
        
        var pluralMinutes = appDelegate.longPauseMinutes == 1 ? NSLocalizedString("minute", comment: "") : NSLocalizedString("minutes", comment: "");
            
        lblSoundsName.text = "\(appDelegate.blindChangeSoundName)/\(appDelegate.lastMinuteSoundName)";
        lblLongPauseMinutes.text = appDelegate.longPauseMinutes == 0 ? NSLocalizedString("Disabled", comment: "") : "\(appDelegate.longPauseMinutes) \(pluralMinutes)";
        lblBackgroundMode.text = appDelegate.backgroundMode == "static" ? NSLocalizedString("Static", comment: "") : NSLocalizedString("Dynamic", comment: "");

        var languageList = NSArray(contentsOfFile: NSBundle.mainBundle().pathForResource("languages", ofType: "plist")!);
        for language : NSDictionary in languageList as! Array<NSDictionary> {
            
            if( (language["culture"] as! String) == appDelegate.language ){
                
                lblLanguage.text = (language["language"] as! String);
                break;
            }
        }
        
        if( Constants.DeviceIdiom.IS_IPAD ){
            
            println("appDelegate.multipeerDeviceName: \(appDelegate.multipeerDeviceName)");
            lblMultipeerDevice.text = appDelegate.multipeerDeviceName;
        }
        
        tableView.reloadData();
    }
}